File: app/src/main/java/com/hmproductions/bingo/adapter/RoomsRecyclerAdapter.java
package com.hmproductions.bingo.adapter;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.hmproductions.bingo.R;
import com.hmproductions.bingo.data.Room;

import java.util.ArrayList;
import java.util.Collections;

import static com.hmproductions.bingo.utils.Constants.CLASSIC_TAG;

public class RoomsRecyclerAdapter extends RecyclerView.Adapter<RoomsRecyclerAdapter.RoomViewHolder> {

    private static final int HEADER_TYPE = -435;
    private static final int NORMAL_TYPE = -107;

    private Context context;
    private final OnRoomItemClickListener listener;
    private ArrayList<Room> roomArrayList;

    public interface OnRoomItemClickListener {
        void onRoomClick(RoomViewHolder viewHolder, int position);
    }

    public RoomsRecyclerAdapter(Context context, ArrayList<Room> roomArrayList, OnRoomItemClickListener listener) {
        this.context = context;
        this.listener = listener;
        this.roomArrayList = roomArrayList;

        createHeadersAndRefactorRooms();
    }

    @NonNull
    @Override
    public RoomViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new RoomViewHolder(LayoutInflater.from(context).inflate(
                viewType == NORMAL_TYPE ? R.layout.room_list_item : R.layout.room_heading_item, parent, false), viewType);
    }

    @Override
    public void onBindViewHolder(@NonNull RoomViewHolder holder, int position) {

        Room currentRoom = roomArrayList.get(position);

        holder.roomNameTextView.setText(currentRoom.getName());

        if (holder.getItemViewType() == NORMAL_TYPE) {
            holder.countTextView.setText(String.valueOf(currentRoom.getCount()));
            holder.maxCountTextView.setText(String.valueOf(currentRoom.getMaxSize()));
        }
    }

    @Override
    public int getItemCount() {
        if (roomArrayList == null || roomArrayList.size() == 0) return 0;
        return roomArrayList.size();
    }

    @Override
    public int getItemViewType(int position) {
        return roomArrayList.get(position).getRoomId() == HEADER_TYPE ? HEADER_TYPE : NORMAL_TYPE;
    }

    public void swapData(ArrayList<Room> newList) {
        this.roomArrayList = newList;
        createHeadersAndRefactorRooms();
        notifyDataSetChanged();
    }

    private void createHeadersAndRefactorRooms() {
        if (roomArrayList != null && roomArrayList.size() > 0) {
            Collections.sort(roomArrayList, (roomA, roomB) -> Integer.compare(
                    roomB.getMaxSize() - roomB.getCount(), roomA.getMaxSize() - roomA.getCount()));

            if (roomArrayList.get(0).getMaxSize() != roomArrayList.get(0).getCount())
                roomArrayList.add(0, new Room(HEADER_TYPE, 0, 0, "Available Rooms"));

            int roomsFullPosition = -1;
            for (Room room : roomArrayList)
                if (room.getMaxSize() == room.getCount() && room.getRoomId() != HEADER_TYPE) {
                    roomsFullPosition = roomArrayList.indexOf(room);
                    break;
                }

            if (roomsFullPosition != -1) {
                roomArrayList.add(roomsFullPosition, new Room(HEADER_TYPE, 0, 0, "Full Rooms"));
            }

            for (Room room : roomArrayList)
                Log.v(CLASSIC_TAG, room.getName() + " == " + room.getCount() + "/" + room.getMaxSize());
        }
    }

    public class RoomViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        TextView roomNameTextView, countTextView, maxCountTextView;

        RoomViewHolder(View itemView, int viewType) {
            super(itemView);

            roomNameTextView = itemView.findViewById(R.id.roomName_textView);

            if (viewType == NORMAL_TYPE) {
                countTextView = itemView.findViewById(R.id.count_textView);
                maxCountTextView = itemView.findViewById(R.id.maxCount_textView);
                itemView.setOnClickListener(this);
            }
        }

        @Override
        public void onClick(View view) {
            listener.onRoomClick(this, getAdapterPosition());
        }
    }

    /* private void test() {
     *   //Mock rooms list for testing
     *   ArrayList<Room> mockList = new ArrayList<>();
     *   mockList.add(new Room(1, 3, 3, "hello1"));
     *   mockList.add(new Room(1, 2, 3, "hello2"));
     *   mockList.add(new Room(1, 0, 3, "hello4"));
     *   mockList.add(new Room(1, 4, 4, "hello5"));
     *   mockList.add(new Room(1, 1, 1, "hello7"));
     *   mockList.add(new Room(1, 4, 6, "hello8"));
     *   mockList.add(new Room(1, 8, 8, "hello9"));
     *   mockList = createHeadersAndRefactorRooms(mockList);
     *   for (Room room : mockList)
     *       Log.v(CLASSIC_TAG, room.getName() + " == " + room.getCount() + "/" + room.getMaxSize());
     }
        */
}

File: app/src/main/java/com/hmproductions/bingo/adapter/RoomsRecyclerAdapter.kt
package com.hmproductions.bingo.adapter

import android.app.Activity
import android.content.Context
import android.graphics.Color
import android.graphics.drawable.GradientDrawable
import android.support.v7.widget.RecyclerView
import android.util.DisplayMetrics
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.RelativeLayout
import com.hmproductions.bingo.R
import com.hmproductions.bingo.data.Room
import com.hmproductions.bingo.utils.Miscellaneous.getTimeLimitString
import com.hmproductions.bingo.utils.TimeLimitUtils
import kotlinx.android.synthetic.main.room_list_item.view.*
import java.util.ArrayList
import kotlin.Comparator

class RoomsRecyclerAdapter(private val context: Context, private var roomArrayList: ArrayList<Room>, private val listener: OnRoomItemClickListener) : RecyclerView.Adapter<RoomsRecyclerAdapter.RoomViewHolder>() {

    companion object {
        private const val HEADER_TYPE = -435
        private const val NORMAL_TYPE = -107
    }

    interface OnRoomItemClickListener {
        fun onRoomClick(view: View, position: Int)
    }

    init {
        createHeadersAndRefactorRooms()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RoomViewHolder = RoomViewHolder(LayoutInflater.from(context).inflate(
            if (viewType == NORMAL_TYPE) R.layout.room_list_item else R.layout.room_heading_item, parent, false), listener)

    override fun onBindViewHolder(holder: RoomViewHolder, position: Int) = holder.bindRoom(roomArrayList[position])

    override fun getItemCount() = roomArrayList.size

    override fun getItemViewType(position: Int) = if (roomArrayList[position].roomId == HEADER_TYPE) HEADER_TYPE else NORMAL_TYPE

    private fun createHeadersAndRefactorRooms() {
        if (roomArrayList.size > 0) {
            roomArrayList.sortWith(Comparator { (_, countA, maxSizeA), (_, countB, maxSizeB) -> Integer.compare(maxSizeB - countB, maxSizeA - countA) })

            if (roomArrayList[0].maxSize != roomArrayList[0].count)
                roomArrayList.add(0, Room(HEADER_TYPE, 0, 0, "Available Rooms", TimeLimitUtils.TIME_LIMIT.MINUTE_1))

            var roomsFullPosition = -1
            for (room in roomArrayList)
                if (room.maxSize == room.count && room.roomId != HEADER_TYPE) {
                    roomsFullPosition = roomArrayList.indexOf(room)
                    break
                }

            if (roomsFullPosition != -1) {
                roomArrayList.add(roomsFullPosition, Room(HEADER_TYPE, 0, 0, "Full Rooms - In game", TimeLimitUtils.TIME_LIMIT.MINUTE_1))
            }
        }
    }

    class RoomViewHolder(itemView: View, private val listener: OnRoomItemClickListener) : RecyclerView.ViewHolder(itemView) {

        fun bindRoom(room: Room?) {
            itemView.roomNameTextView.text = room?.name

            if (itemViewType == NORMAL_TYPE) {
                with(itemView) {
                    countTextView.text = "${room?.count}"
                    maxCountTextView.text = "${room?.maxSize}"
                    timeLimitTextView.text = getTimeLimitString(room?.timeLimit)
                    roomIconTextView.text = room?.name?.get(0).toString().toUpperCase()

                    val displayMetrics = DisplayMetrics()
                    (this.context as Activity).windowManager.defaultDisplay.getMetrics(displayMetrics)

                    if(displayMetrics.widthPixels < 500) {
                        val params = roomIconTextView.layoutParams as RelativeLayout.LayoutParams
                        params.height = context.resources.getDimensionPixelSize(R.dimen.small_room_icon)
                        params.width = context.resources.getDimensionPixelSize(R.dimen.small_room_icon)
                        roomIconTextView.layoutParams = params

                        roomNameTextView.setTextSize(TypedValue.COMPLEX_UNIT_SP, 16F)
                        timeLimitTextView.setTextSize(TypedValue.COMPLEX_UNIT_SP, 16F)
                        roomIconTextView.setTextSize(TypedValue.COMPLEX_UNIT_SP, 16F)
                    }

                    roomIconTextView.setBackgroundResource(R.drawable.cell_circle_foreground)
                    (roomIconTextView.background as GradientDrawable).setColor(Color.parseColor("#EEEEEE"))
                    setOnClickListener { listener.onRoomClick(itemView, adapterPosition) }
                }
            }
        }
    }
}